File: 155-min-stack/min-stack.cpp
class MinStack {
public:
    stack<long long> st;
    long long mini;

    MinStack() {
        mini = LLONG_MAX; 
    }

    void push(int val) {
        long long x = val;

        if (st.empty()) {
            mini = x;
            st.push(x);
        } else {
            if (x < mini) {
                st.push(2LL * x - mini); 
                mini = x;
            } else {
                st.push(x);
            }
        }
    }

    void pop() {
        if (st.empty())
            return;

        long long el = st.top();
        st.pop();

        if (el < mini) {
            mini = 2 * mini - el; 
        }
    }

    int top() {
        if (st.empty())
            return -1;

        long long el = st.top();
        return (el < mini) ? mini : el;
    }

    int getMin() {
        return mini;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.st1 = []
        self.st2 = []
        

    def push(self, val: int) -> None:
        self.st1.append(val)
        curr_min = self.st1[-1]
        for i in self.st1:
            if i < curr_min:
                curr_min = i
        self.st2.append(curr_min) 

        

    def pop(self) -> None:
        self.st1.pop()
        self.st2.pop()
        

    def top(self) -> int:
        return self.st1[-1]
        

    def getMin(self) -> int:
        return self.st2[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()